pkg/routeagent_driver/handlers/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	if h.config.MaxPacketLossCount != 0 {
+		pingerConfig.MaxPacketLossCount = h.config.MaxPacketLossCount
+	}
+
+	newPingerFunc := h.config.NewPinger
+	if newPingerFunc == nil {
+		newPingerFunc = pinger.NewPinger
+	}
+
+	pingerObject := newPingerFunc(pingerConfig)
+	h.pingers[endpoint.Spec.CableName] = pingerObject
+	pingerObject.Start()
+
+	logger.Infof("HealthChecker started pinger for CableName: %q with HealthCheckIP %q",
+		endpoint.Spec.CableName, endpoint.Spec.HealthCheckIP)
+
+	return nil
+}
+
+func (h *controller) RemoteEndpointDeleted(endpoint *submarinerv1.Endpoint) error {
+	h.Lock()
+	defer h.Unlock()
+
+	if pingerObject, found := h.pingers[endpoint.Spec.CableName]; found {
+		pingerObject.Stop()
+		delete(h.pingers, endpoint.Spec.CableName)
+	}
+
+	return nil
+}
+
+func (h *controller) Init() error {
+	go func() {
+		wait.Until(func() {
+			h.Lock()
+			defer h.Unlock()
+
+			h.syncRouteAgentStatus()
+		}, RouteAgentUpdateInterval, h.stopCh)
+	}()
+
+	return nil
+}
+
+// TransitionToNonGateway is called once for each transition of the local node from Gateway to a non-Gateway.
+func (h *controller) TransitionToNonGateway() error {
+	h.Lock()
+	defer h.Unlock()
+
+	if h.config.HealthCheckerEnabled {
+		remoteEndpoints := h.State().GetRemoteEndpoints()
+
+		for i := range remoteEndpoints {
+			err := h.processEndpointCreatedOrUpdated(&remoteEndpoints[i])
+			if err != nil {
+				return errors.Wrapf(err, "error processing remote endpoint %s", remoteEndpoints[i].Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// TransitionToGateway is called once for each transition of the local node from non-Gateway to a Gateway.
+func (h *controller) TransitionToGateway() error {
+	h.Lock()
+	defer h.Unlock()
+
+	if h.config.HealthCheckerEnabled {
+		for i := range h.pingers {
+			h.pingers[i].Stop()
+			delete(h.pingers, i)
+		}
+	}
+
+	h.syncRouteAgentStatus()
+
+	return nil
+}
+
+func (h *controller) GetNetworkPlugins() []string {
+	return []string{event.AnyNetworkPlugin}
+}
+
+func (h *controller) GetName() string {
+	return "routeAgent-health-checker"
+}
+
+func (h *controller) syncRouteAgentStatus() {
+	routeAgent := h.generateRouteAgentObject()
+	remoteEndpoints := h.State().GetRemoteEndpoints()
+
+	for i := range remoteEndpoints {
+		var connectionStatus submarinerv1.ConnectionStatus
+		var remoteEndpoint submarinerv1.RemoteEndpoint
+		var statusMessage string
+		var latencyRTT *submarinerv1.LatencyRTTSpec
+
+		if !h.config.HealthCheckerEnabled {
+			connectionStatus = submarinerv1.ConnectionNone
+			statusMessage = "Health check is not enabled"
+		} else if h.State().IsOnGateway() {
+			connectionStatus = submarinerv1.ConnectionNone
+			statusMessage = "Health check is not performed on gateway nodes"
+		} else {
+			pingerObject, found := h.pingers[remoteEndpoints[i].Spec.CableName]
+			if !found {
+				logger.Warningf("Pinger not found for %q", remoteEndpoints[i].Spec.CableName)
+				continue
+			}
```

Comment:
I don't think we need a warning here. This could normally happen in the small window between when a `RemoteEndpoint` is added to the cache and the handler is invoked. The next sync period would pick it up. Either way we can still add it with status `submarinerv1.Connecting`.  That's why I initialized `remoteEndpoint` before this block [here](https://github.com/submariner-io/submariner/pull/3077#discussion_r1724161667).  It ensures `Status`  is always set and eliminates the need for the else on L276.

```suggestion
		} else if pingerObject, found := h.pingers[remoteEndpoints[i].Spec.CableName]; found {
			pingerObject, found := h.pingers[remoteEndpoints[i].Spec.CableName]
```

Reply:
Done.